frontend-service/src/test/java/com/amazonaws/blox/frontend/mappers/StartDeploymentMapperTest.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://aws.amazon.com/apache2.0/
+ *
+ * or in the "LICENSE" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazonaws.blox.frontend.mappers;
+
+import com.amazonaws.blox.dataservicemodel.v1.model.EnvironmentId;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.StartDeploymentRequest;
+import com.amazonaws.blox.dataservicemodel.v1.model.wrappers.StartDeploymentResponse;
+import com.amazonaws.blox.frontend.operations.StartDeployment;
+import com.amazonaws.serverless.proxy.internal.model.ApiGatewayRequestContext;
+import org.junit.Before;
+import org.junit.Test;
+import org.mapstruct.factory.Mappers;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class StartDeploymentMapperTest {
+
+  private static final String ACCOUNT_ID = "123456789012";
+  private static final String CLUSTER = "mycluster";
+  private static final String ENVIRONMENT_NAME = "mynev";
+  private static final String ENVIRONMENT_REVISION_ID = "1";
+  private static final String DEPLOYMENT_ID = "deployment-id";
+
+  private final StartDeploymentMapper mapper = Mappers.getMapper(StartDeploymentMapper.class);
```

Comment:
I wonder if it might be a bit more straightforward to test the mapper implicitly from the StartDeploymentTest, and to then avoid mocking out the mapper there. Thoughts?

Reply:
I think it's what we want to discuss f-2-f.  Just summarize what I think about them:

### Unit Test Practice 1:  API Unit test with real Mapper 
Example see (https://github.com/blox/blox/blob/dev/frontend-service/src/test/java/com/amazonaws/blox/frontend/operations/CreateEnvironmentTest.java).  It will use a real mapper inside API unit test. 

I can see this unit test can gain more confidence it pass the real object and get some concrete result to check.  However, the downside I see are:
 - Not a pure unit test for API test only. It actually depends on the implementation of mapper. If any mapper model change or rule change, it most likely we need to fix the API test even no change on API class. It validates one of the rule that one bug should be only be caught in one test.
- No Mapper unit test:  Most likely we think we exercise the mapper in API test, but API unit test is not focus on the mapper test. Without mapper unit test, we cannot make sure we have cover all the cases in the mapper. 

### Unit Test Practice 2:  Unit test for mapper and Mock mapper/object in API Unit test
Example see:
 - Mapper unit test: https://github.com/blox/blox/blob/dev/frontend-service/src/test/java/com/amazonaws/blox/frontend/mappers/StartDeploymentMapperTest.java
 - API unit test: https://github.com/blox/blox/blob/dev/frontend-service/src/test/java/com/amazonaws/blox/frontend/operations/StartDeploymentTest.java

The good part is testing each class separately without any dependency.  The argument is always goes to:
 - Mapper consider as an implementation detail in API. If any refactoring on mapper, say change API name from one to other.  We don't need to update the test, and it should expect the same result.  (The practice 1 may argue that how would you make sure you call the right mapper API, what if you make a call to the wrong API but just got right result )

It's kind of the argument on __Testing Behavior__  vs. __Testing Implementation__.   

The first test practice is more on testing behavior, the test says "I don't care about what is mapper,  with this input and I expect this result."

The second test practice is on testing implementation, the test says "This class calls this API of mapper, and then calls data service with the result from mapper."    

It's may a personal taste thing, but I'm open to have a discussion.